### PR TITLE
Various fixes for execdriver

### LINKED ISSUE #11764: Container termination events lost in 1.6.0
### Summary:

After upgrading to docker 1.6.0, containers cannot be destroyed. The containers run as normal, but the parent process never perceives the container as exited. For example, if I run `docker run --entrypoint=/bin/echo busybox test`, it prints `test` and then hangs indefinitely.

Attempting to `kill`, `stop`, or `rm -f` a container results in:

```
Error response from daemon: Cannot kill container 52e: [2] Container does not exist: container destroyed
FATA[0000] Error: failed to kill one or more containers
```

This only happens on one of the two servers I've upgraded to `1.6.0-rc2` (both of which have the same output for the "System" section below). It happens consistently with any combination of parameters to `docker run`. This was also the behaviour with `1.6.0-rc1`.

I've attached `docker.log` captures for the whole process and attached them to this issue.

I can capture `strace`s or whatever other instrumentation seems helpful, just ask if you need more, or have some pointers on how I might fix this myself.

### System:

```
$ uname -r
3.18.8-031808-generic
$ docker --version
Docker version 1.6.0-rc2, build c5ee149
$ cat /etc/lsb-release | grep DESC
DISTRIB_DESCRIPTION="Ubuntu 14.04 LTS"
```

### Docker logs

1) start docker

```
INFO[0000] +job serveapi(unix:///var/run/docker.sock)
INFO[0000] Listening for HTTP on unix (/var/run/docker.sock)
INFO[0000] +job init_networkdriver()
INFO[0000] -job init_networkdriver() = OK (0)
WARN[0000] mountpoint for memory not found
INFO[0000] Loading containers: start.
................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
INFO[0000] Loading containers: done.
INFO[0000] docker daemon: 1.6.0-rc2 c5ee149; execdriver: native-0.2; graphdriver: overlay
INFO[0000] +job acceptconnections()
INFO[0000] -job acceptconnections() = OK (0)
INFO[0000] Daemon has completed initialization
INFO[0002] GET /v1.18/containers/json
INFO[0002] +job containers()
INFO[0002] -job containers() = OK (0)
INFO[0004] GET /v1.18/containers/json
INFO[0004] +job containers()
INFO[0004] -job containers() = OK (0)
INFO[0009] GET /v1.18/containers/json
INFO[0009] +job containers()
INFO[0009] -job containers() = OK (0)
INFO[0013] GET /v1.18/containers/json
INFO[0013] +job containers()
INFO[0014] -job containers() = OK (0)
INFO[0014] GET /v1.18/containers/json
INFO[0014] +job containers()
INFO[0014] -job containers() = OK (0)
```

2) `docker run --entrypoint=/bin/echo busybox test`

```
INFO[0014] POST /v1.18/containers/create
INFO[0014] +job create()
INFO[0014] +job log(create, 52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8, busybox:latest)
INFO[0014] -job log(create, 52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8, busybox:latest) = OK (0)
INFO[0014] -job create() = OK (0)
INFO[0014] POST /v1.18/containers/52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8/attach?stderr=1&stdout=1&stream=1
INFO[0014] +job container_inspect(52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8)
INFO[0014] -job container_inspect(52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8) = OK (0)
INFO[0014] +job attach(52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8)
INFO[0014] POST /v1.18/containers/52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8/start
INFO[0014] +job start(52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8)
INFO[0014] +job allocate_interface(52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8)
INFO[0014] -job allocate_interface(52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8) = OK (0)
INFO[0014] +job log(start, 52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8, busybox:latest)
INFO[0014] -job log(start, 52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8, busybox:latest) = OK (0)
INFO[0014] GET /images/json?all=False
INFO[0014] +job images()
INFO[0014] -job images() = OK (0)
INFO[0014] GET /images/json?all=True
INFO[0014] +job images()
INFO[0014] -job images() = OK (0)
INFO[0015] GET /containers/json?all=False&size=False
INFO[0015] +job containers()
INFO[0015] -job start(52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8) = OK (0)
INFO[0015] -job containers() = OK (0)
WARN[0015] Your kernel does not support OOM notifications: There is no path for "memory" in state
INFO[0015] GET /containers/json?all=True&size=False
INFO[0015] +job containers()
INFO[0015] -job containers() = OK (0)
INFO[0015] GET /events?since=1427317521&until=1427317579
INFO[0015] +job events()
INFO[0015] -job events() = OK (0)
INFO[0018] GET /v1.18/containers/json
INFO[0018] +job containers()
INFO[0018] -job containers() = OK (0)
INFO[0020] GET /v1.18/containers/json
INFO[0020] +job containers()
INFO[0020] -job containers() = OK (0)
INFO[0022] GET /v1.18/containers/json
INFO[0022] +job containers()
INFO[0022] -job containers() = OK (0)
INFO[0025] GET /v1.18/containers/json
INFO[0025] +job containers()
INFO[0025] -job containers() = OK (0)
```

3) Notice that container hasn't terminated yet. Hit `^C` in shell running container.

```
INFO[0026] POST /v1.18/containers/52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8/kill?signal=INT
INFO[0026] +job kill(52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8, INT)
Cannot kill container 52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8: [2] Container does not exist: container destroyed
INFO[0026] -job kill(52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8, INT) = ERR (1)
ERRO[0026] Handler for POST /containers/{name:.*}/kill returned error: Cannot kill container 52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8: [2] Container does not exist: container destroyed
ERRO[0026] HTTP Error: statusCode=500 Cannot kill container 52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8: [2] Container does not exist: container destroyed
INFO[0029] GET /v1.18/containers/json
INFO[0029] +job containers()
INFO[0029] -job containers() = OK (0)
INFO[0032] GET /v1.18/containers/json
INFO[0032] +job containers()
INFO[0032] -job containers() = OK (0)
INFO[0033] POST /v1.18/containers/52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8/kill?signal=HUP
INFO[0033] +job kill(52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8, HUP)
Cannot kill container 52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8: [2] Container does not exist: container destroyed
INFO[0033] -job kill(52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8, HUP) = ERR (1)
ERRO[0033] Handler for POST /containers/{name:.*}/kill returned error: Cannot kill container 52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8: [2] Container does not exist: container destroyed
ERRO[0033] HTTP Error: statusCode=500 Cannot kill container 52e73e5961d5d5a76a5a70e9a9e253af6dc1b5cfa95e8623cc820f46eae52ca8: [2] Container does not exist: container destroyed
INFO[0033] GET /v1.18/containers/json
INFO[0033] +job containers()
INFO[0033] -job containers() = OK (0)
INFO[0034] GET /images/json?all=False
INFO[0034] +job images()
INFO[0034] -job images() = OK (0)
INFO[0034] GET /images/json?all=True
INFO[0034] +job images()
INFO[0034] -job images() = OK (0)
INFO[0034] GET /containers/json?all=False&size=False
INFO[0034] +job containers()
INFO[0034] -job containers() = OK (0)
INFO[0034] GET /containers/json?all=True&size=False
INFO[0034] +job containers()
INFO[0034] -job containers() = OK (0)
INFO[0034] GET /events?since=1427317579&until=1427317598
INFO[0034] +job events()
INFO[0034] -job events() = OK (0)
INFO[0036] GET /v1.18/containers/json
INFO[0036] +job containers()
INFO[0036] -job containers() = OK (0)
```

### PR DESCRIPTION
**Pull request by LK4D4**
Fix #11764

### CODE REVIEW

File: daemon/execdriver/native/driver.go [hunk 1]
```diff
@@ -162,18 +162,18 @@ func (d *driver) Run(c *execdriver.Command, pipes *execdriver.Pipes, startCallba
...
 		// we need such hack for tracking processes with inerited fds,
 		// because cmd.Wait() waiting for all streams to be copied
 		waitF = waitInPIDHost(p, cont)
 	}
 	ps, err := waitF()
 	if err != nil {
-		if err, ok := err.(*exec.ExitError); !ok {
+		execErr, ok := err.(*exec.ExitError)
+		if !ok {
 			return execdriver.ExitStatus{ExitCode: -1}, err
```
Comment (crosbymichael) [inline daemon/execdriver/native/driver.go]: Wound't returning here miss the container.Destroy call?

Comment (LK4D4) [inline daemon/execdriver/native/driver.go]: There is spare in defer.

**Comment by crosbymichael** [comment]
LGTM

**Comment by jfrazelle** [comment]
LGTM

**Comment by jfrazelle** [comment]
hang on running lxc on this

**Comment by jfrazelle** [comment]
LGTM

**Pull request merged**